### ISSUE TITLE
AttributeError: '_thread._local' object has no attribute 'Hub'

### ISSUE THREAD

**Issue by sibelius** [eventlet/eventlet#325]
I'm having this issue in python 3.5

```python
Exception in thread pymongo_server_monitor_thread:
Traceback (most recent call last):
  File "/app/lib/python3.5/site-packages/eventlet/hubs/__init__.py", line 111, in get_hub
    hub = _threadlocal.hub
AttributeError: '_thread._local' object has no attribute 'hub'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/app/lib/python3.5/site-packages/eventlet/hubs/__init__.py", line 114, in get_hub
    _threadlocal.Hub
AttributeError: '_thread._local' object has no attribute 'Hub'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 914, in _bootstrap_inner
    self.run()
  File "/usr/local/Cellar/python3/3.5.1/Frameworks/Python.framework/Versions/3.5/lib/python3.5/threading.py", line 862, in run
    self._target(*self._args, **self._kwargs)
  File "/app/lib/python3.5/site-packages/pymongo/periodic_executor.py", line 103, in _run
    time.sleep(self._min_interval)
  File "/app/lib/python3.5/site-packages/eventlet/greenthread.py", line 29, in sleep
    hub = hubs.get_hub()
  File "/app/lib/python3.5/site-packages/eventlet/hubs/__init__.py", line 116, in get_hub
    use_hub()
  File "/app/lib/python3.5/site-packages/eventlet/hubs/__init__.py", line 70, in use_hub
    mod = get_default_hub()
  File "/app/lib/python3.5/site-packages/eventlet/hubs/__init__.py", line 39, in get_default_hub
    return eventlet.hubs.epolls
AttributeError: module 'eventlet.hubs' has no attribute 'epolls'
```

**Comment by sibelius** [comment]
this is not working on python 3.4 as well

**Comment by sibelius** [comment]
maybe this help debug this:

```python
def patch_eventlets():
    import eventlet
    eventlet.monkey_patch()
    import eventlet.green.ssl

    def _green_ssl_recv_into (self, buffer, nbytes=None, flags=0):
        if self._sslobj:
            if flags != 0:
                raise ValueError(
                    "non-zero flags not allowed in calls to recv_into() on %s" %
                    self.__class__)
            if nbytes is None:
                if buffer:
                    nbytes = len(buffer)
                else:
                    nbytes = 1024
            read = self.read(nbytes, buffer)
            return read
        else:
            while True:
                try:
                    return eventlet.green.ssl.socket.recv_into(self, buffer, nbytes, flags)
                except eventlet.green.ssl.orig_socket.error as e:
                    if self.act_non_blocking:
                        raise
                    erno = eventlet.green.ssl.get_errno(e)
                    if erno in eventlet.green.ssl.greenio.SOCKET_BLOCKING:
                        try:
                            eventlet.green.ssl.trampoline(
                                self, read=True,
                                timeout=self.gettimeout(), timeout_exc=eventlet.green.ssl.timeout_exc('timed out'))
                        except eventlet.green.ssl.IOClosed:
                            return b''
                    elif erno in eventlet.green.ssl.greenio.SOCKET_CLOSED:
                        return b''
                    raise

    eventlet.green.ssl.GreenSSLSocket.recv_into = _green_ssl_recv_into
```

I use this on my celery worker

**Comment by sibelius** [comment]
it works on linux, but no on my mac os x (el captain)

**Comment by rickwargo** [comment]
epoll is not supported on OS X; you'll have to use kqueue instead.

**Comment by temoto** [comment]
This is a surprise. Please, show

```
import select
import os
print('env:', os.environ)
print('select.epoll:', repr(getattr(select, 'epoll')))
```

**Comment by rickwargo** [comment]
I was just researching how to get epoll on OS X and found these: [https://www.google.com/search?q=epoll+osx](https://www.google.com/search?q=epoll+osx)

```
rick$ ipython
Python 3.5.2 (default, Jun 29 2016, 13:43:58)
Type "copyright", "credits" or "license" for more information.

IPython 5.1.0 -- An enhanced Interactive Python.
?         -> Introduction and overview of IPython's features.
%quickref -> Quick reference.
help      -> Python's own help system.
object?   -> Details about 'object', use 'object??' for extra details.

In [1]: import select
   ...: import os
   ...: print('env:', os.environ)
   ...: print('select.epoll:', repr(getattr(select, 'epoll')))
   ...:
env: environ({'AWS_DEFAULT_PROFILE': 'CLI', 'DISPLAY': '/private/tmp/com.apple.launchd.fw87cd2qEE/org.macosforge.xquartz:0', 'TERM': 'xterm-256color', 'USER': 'rick', 'ITERM_PROFILE': 'Default', 'SHELL': '/bin/bash', 'SHLVL': '1', '__PYVENV_LAUNCHER__': '/usr/local/Cellar/python3/3.5.2/bin/python3.5', 'SSH_AUTH_SOCK': '/private/tmp/com.apple.launchd.aowOyszqGu/Listeners', 'PATH': '/Library/Frameworks/Python.framework/Versions/2.7/bin:/Users/rick/.rbenv/shims:/usr/local/bin:/usr/bin:/bin:/usr/sbin:/sbin:/opt/X11/bin', 'TMPDIR': '/var/folders/vq/2p4wvq650mscsmtm94zb3qtm0000gn/T/', 'Apple_PubSub_Socket_Render': '/private/tmp/com.apple.launchd.GHr6To1jXk/Render', 'XPC_SERVICE_NAME': '0', '__CF_USER_TEXT_ENCODING': '0x1F5:0x0:0x0', 'TERM_PROGRAM_VERSION': '3.0.7', 'HOME': '/Users/rick', 'XPC_FLAGS': '0x0', '_': '/usr/local/bin/ipython', 'ITERM_SESSION_ID': 'w1t3p1:D9950A19-1689-4F85-9998-94D27344BE8C', 'LANG': 'en_US.UTF-8', 'TERM_SESSION_ID': 'w1t3p1:D9950A19-1689-4F85-9998-94D27344BE8C', 'PWD': '/Users/rick/Code/Python/CeleryLocalTest', 'TERM_PROGRAM': 'iTerm.app', 'LOGNAME': 'rick'})
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-1-d19933734f45> in <module>()
      2 import os
      3 print('env:', os.environ)
----> 4 print('select.epoll:', repr(getattr(select, 'epoll')))

AttributeError: module 'select' has no attribute 'epoll'
```

**Comment by temoto** [comment]
@rickwargo yup that's how it should behave and seemingly @sibelius has something else.

You are not supposed to run epoll on OSX, Eventlet must choose kqueue or select automatically.

**Comment by uclatommy** [comment]
I'm getting the same issue with the same lines in my call stack. If you look at the `get_default_hub()` function, it tries to auto-detect which hub it should use:
```
    try:
        import eventlet.hubs.epolls
        return eventlet.hubs.epolls
    except ImportError:
        try:
            import eventlet.hubs.kqueue
            return eventlet.hubs.kqueue
```
But this results in the `AttributeError: module 'eventlet.hubs' has no attribute 'epolls'` error.

So the failure happens because the `except` line doesn't catch an `AttributeError`, which is what gets thrown when you try to import `eventlet.hubs.epolls`.

If you look one line higher in the call stack, it looks like there's a way you can force your system to select `kqueue` instead of letting eventlet run its `get_default_hub()` function. You can see this in `use_hub()`:
```
    if mod is None:
        mod = os.environ.get('EVENTLET_HUB', None)
    if mod is None:
        mod = get_default_hub()
```

TLDR; set your environment variable `EVENTLET_HUB=eventlet.hubs.kqueue` to force eventlet to choose `kqueue` and avoid this error.

**Comment by temoto** [comment]
Hub import mechanism was improved, try latest version, it will reveal root
cause.

**Comment by uclatommy** [comment]
Thanks for the advice. Just wanted to report back that the issue persisted even after I set my environment variable. However, the error disappeared after I installed from master.

Is there any consensus on when the next release will be? I'd prefer to be using a release version, but I'll stick with the current head revision until then.

**Comment by macdonjo** [comment]
I also have this error with eventlet 0.24.1.
